application/src/main/java/org/opentripplanner/routing/algorithm/mapping/ViaLocationTypeMapper.java

```diff
@@ -0,0 +1,53 @@
+package org.opentripplanner.routing.algorithm.mapping;
+
+import java.util.Objects;
+import javax.annotation.Nullable;
+import org.opentripplanner.model.plan.leg.ViaLocationType;
+import org.opentripplanner.routing.api.request.RouteRequest;
+import org.opentripplanner.street.model.vertex.TemporaryStreetLocation;
+import org.opentripplanner.transit.model.site.StopLocation;
+
+/**
+ * Maps locations to {@link ViaLocationType} based on their existence in the route request.
+ */
+public class ViaLocationTypeMapper {
+
+  @Nullable
+  public static ViaLocationType map(RouteRequest request, StopLocation stop) {
+    return request
+      .listViaLocations()
+      .stream()
+      .flatMap(viaLocation ->
+        viaLocation
+          .stopLocationIds()
+          .stream()
+          .map(stopId -> {
+            // This might yield to false positive matches (stop location is visited multiple times
+            // in an itinerary), but those cases should be quite rare.
+            if (stopId.equals(stop.getId()) || stopId.equals(stop.getStationOrStopId())) {
```

Comment:
This also misses Group-of-stops and multi-module-stops, and fail to map the right via point the right stop if the via-stop appears multiple times. This method needs to now the via sequence number to do it correct.

This is business logic and should not be in a mapper. However, this follows the established pattern. This is not the most critical design mistake in OTP, so I am tempted to let it pass an focus on improving the more important parts.

Reply:
Same as above.